File: packages/Recognizers/ConversationLanguageUnderstanding/dotnet/CluRecognizerOptionsBase.cs
using System;
using System.Net.Http;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.Bot.Builder;
using Microsoft.Bot.Builder.Dialogs;
using Microsoft.Bot.Schema;
using Microsoft.Extensions.Options;

namespace Microsoft.Bot.Components.Recognizers.CLURecognizer
{
    /// <summary>
    /// CLU Recognizer Options.
    /// </summary>
    public abstract class CluRecognizerOptionsBase
    {
        /// <summary>
        /// Initializes a new instance of the <see cref="CluRecognizerOptionsBase"/> class.
        /// </summary>
        /// <param name="application">An instance of <see cref="CluApplication"/>.</param>
        protected CluRecognizerOptionsBase(CluApplication application)
        {
            Application = application ?? throw new ArgumentNullException(nameof(application));
        }

        /// <summary>
        /// Gets the CLU application used to recognize text.
        /// </summary>
        /// <value>
        /// The CLU application to use to recognize text.
        /// </value>
        public CluApplication Application { get; private set; }

        /// <summary>
        /// Gets or sets the time in milliseconds to wait before the request times out.
        /// </summary>
        /// <value>
        /// The time in milliseconds to wait before the request times out. Default is 100000 milliseconds.
        /// </value>
        public double Timeout { get; set; } = CluConstants.HttpClientOptions.Timeout;

        /// <summary>
        /// Gets or sets the IBotTelemetryClient used to log the CluResult event.
        /// </summary>
        /// <value>
        /// The client used to log telemetry events.
        /// </value>
        public IBotTelemetryClient TelemetryClient { get; set; } = new NullBotTelemetryClient();

        /// <summary>
        /// Gets or sets a value indicating whether to log personal information that came from the user to telemetry.
        /// </summary>
        /// <value>If true, personal information is logged to Telemetry; otherwise the properties will be filtered.</value>
        public bool LogPersonalInformation { get; set; }

        /// <summary>
        /// Gets or sets a value indicating whether flag to indicate if full results from the CLU API should be returned with the recognizer result.
        /// </summary>
        /// <value>A value indicating whether full results from the CLU API should be returned with the recognizer result.</value>
        public bool IncludeAPIResults { get; set; }

        /// <summary>
        /// Gets or sets a value indicating the string index type to include in the the CLU request body.
        /// </summary>
        /// <value>A value indicating the string index type to include in the the CLU request body.</value>
        public string CluRequestBodyStringIndexType { get; set; } = CluConstants.RequestOptions.StringIndexType;

        /// <summary>
        /// Gets or sets a value indicating the api version of the CLU service.
        /// </summary>
        /// <value>A value indicating the api version of the CLU service.</value>
        public string CluApiVersion { get; set; } = CluConstants.RequestOptions.ApiVersion;

        /// <summary>
        /// Gets or sets a value indicating the logical name of the client to create.
        /// </summary>
        /// <value>A value indicating the logical name of the client to create.</value>
        public string LogicalHttpClientName { get; set; } = Options.DefaultName;

        // Support original ITurnContext
        internal abstract Task<RecognizerResult> RecognizeInternalAsync(ITurnContext turnContext, HttpClient httpClient, CancellationToken cancellationToken);

        // Support DialogContext
        internal abstract Task<RecognizerResult> RecognizeInternalAsync(DialogContext context, Activity activity, HttpClient httpClient, CancellationToken cancellationToken);

        // Support string utterance
        internal abstract Task<RecognizerResult> RecognizeInternalAsync(string utterance, HttpClient httpClient, CancellationToken cancellationToken);
    }
}

File: packages/Recognizers/ConversationLanguageUnderstanding/js/src/cluRecognizerOptionsBase.ts
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

import {
  BotTelemetryClient,
  NullTelemetryClient,
  Activity,
  RecognizerResult,
  TurnContext,
} from 'botbuilder';

import { CluApplication } from './cluApplication';
import { CluConstants } from './cluConstants';
import { HttpClient } from '@azure/ms-rest-js';
import { DialogContext } from 'botbuilder-dialogs';

export interface CluRecognizerOptionsBaseFields {
  telemetryClient: BotTelemetryClient;
  logPersonalInformation: boolean;
  includeAPIResults: boolean;
  cluRequestBodyStringIndexType: string;
  cluApiVersion: string;
}

/**
 * CLU Recognizer Options.
 */
export abstract class CluRecognizerOptionsBase {
  private _application!: CluApplication;

  /**
   * Gets or sets the time in milliseconds to wait before the request times out.
   * @returns The time in milliseconds to wait before the request times out. Default is 100000 milliseconds.
   */
  timeout: number = CluConstants.HttpClientOptions.Timeout;

  /**
   * Gets or sets the BotTelemetryClient used to log the CluResult event.
   * @returns The client used to log telemetry events.
   */
  telemetryClient: BotTelemetryClient;

  /**
   * Gets or sets a value indicating whether to log personal information that came from the user to telemetry.
   * @returns If true, personal information is logged to Telemetry; otherwise the properties will be filtered.
   */
  logPersonalInformation: boolean = false;

  /**
   * Gets or sets a value indicating whether flag to indicate if full results from the CLU API should be returned with the recognizer result.
   * @returns A value indicating whether full results from the CLU API should be returned with the recognizer result.
   */
  includeAPIResults: boolean = false;

  /**
   * Gets or sets a value indicating the string index type to include in the the CLU request body.
   * @returns A value indicating the string index type to include in the the CLU request body.
   */
  cluRequestBodyStringIndexType: string;

  /**
   * Gets or sets a value indicating the api version of the CLU service.
   * @returns A value indicating the api version of the CLU service.
   */
  cluApiVersion: string;

  /**
   * Gets the CLU application used to recognize text.
   * @returns The CLU application to use to recognize text.
   */
  get application() {
    return this._application;
  }

  /**
   * Initializes a new instance of the CluRecognizerOptionsBase class.
   * @param application An instance of CluApplication.
   * @param fields The fields to load to the base class.
   */
  protected constructor(
    application: CluApplication,
    fields?: CluRecognizerOptionsBaseFields
  ) {
    if (!application) {
      throw new Error();
    }

    this._application = application;
    this.telemetryClient = fields?.telemetryClient ?? new NullTelemetryClient();
    this.logPersonalInformation = fields?.logPersonalInformation ?? false;
    this.includeAPIResults = fields?.includeAPIResults ?? false;
    this.cluRequestBodyStringIndexType =
      fields?.cluRequestBodyStringIndexType ??
      CluConstants.RequestOptions.StringIndexType;
    this.cluApiVersion =
      fields?.cluApiVersion ?? CluConstants.RequestOptions.ApiVersion;
  }

  abstract recognize(
    turnContext: TurnContext,
    httpClient: HttpClient
  ): Promise<RecognizerResult>;

  abstract recognize(
    dialogContext: DialogContext,
    activity: Activity,
    httpClient: HttpClient
  ): Promise<RecognizerResult>;

  abstract recognize(
    utterance: string,
    httpClient: HttpClient
  ): Promise<RecognizerResult>;
}
